### ISSUE TITLE
Полезные трюки: i-bem

### ISSUE THREAD

**Issue by belozyorcev** [bem/bem-forum-content-ru#270]
Возникла проблема с ```this```. Каждый с ней сталкивался, когда приходилось что-то делать с фильтрами или чем-то другим. Приходилось кэшировать в переменную ```self```
Или например каждый раз при смене модификатора элемента приходилось писать что-то типо этого
```
this.setMod( this.elem('right'), 'size', 'xxxxl');
```

много раз приходится дублировать this... А иногда не сразу понятно на что он ссылается.
При создании одного из блоков я подобрал для себя решение, может кому пригодится.

```
modules.define('scroller', ['i-bem__dom', 'jquery'], function(provide, BEMDOM, $) {

var block;

function elem(name) {
    return block.elem(name);
}

BEMDOM.decl(this.name, {
    onSetMod: {
        'js' : {
            'inited' : function() {
                block = this;
            }
        }
    },
}, {});

provide(BEMDOM);

});

```
После чего к блоку всегда можно обращаться через ```block``` а к элементу через ```elem('name')```.

из плюсов
- однозначность ```block```
- более короткая запись для доступа к элементам ```elem('name')```
- 4 строчки кода

П.Н.
Данное решение актуально для bem-core@v2, в bem-core@v3 - решение будет лучше из коробки.

**Comment by tadatuta** [comment]
@belozyorcev 
Не-не-не-не-не, Девид Блейн! Нельзя так делать!

`this` должен указывать на **экземпляр каждого** блока, а тут получается, что он всегда будет указывать на того, кто последний инициализировался.

Чтобы было понятно о чем речь, добавь в своем примере подписку на клик и положи несколько `scroller` на страницу:

```js
this.bindTo('click', function() {
    console.log(block.domElem); // при клике на любой будет выводить ссылку на последний
});
```

> Каждый с ней сталкивался, когда приходилось что-то делать с фильтрами

JFYI, методы типа `filter()`, `map()` или `forEach()` последним необязательным параметром принимают контекст, как и многие из методов `i-bem.js`.

**Comment by belozyorcev** [comment]
@tadutata

а разве функция не задаёт "замыкание" переменной?

```function(provide, BEMDOM, $) {}```

**Comment by tadatuta** [comment]
В том-то и дело ;)

Тут важно понимать, что модуль в памяти создается один раз, а onSetMod js inited дергается столько раз, сколько блоков окажется на странице. И в замыкании останется ссылка на тот экземпляр блока, который заинитится последним.

**Comment by belozyorcev** [comment]
Эххх ) Жаль... Решение было так близко )

**Comment by belozyorcev** [comment]
@tadutata
Есть ещё один вариант, вдохновлённый bem-core@v3

Что скажешь?

```
modules.define('scroller', ['i-bem__dom', 'jquery'], function(provide, BEMDOM, $) {

BEMDOM.decl(this.name, {
    onSetMod: {
        'js' : {
            'inited' : function() {
                this.el('body').setMod('size', 'xxx');
            }
        }
    },

    el: function (name) {
        var elem = this.elem(name);
        var self = this;
        elem.setMod = function (mod, val) {
            self.setMod(this, mod, val)
        };
        return elem;
    },

}, {});

provide(BEMDOM);

});


```

**Comment by Guria** [comment]
Такую вещь проще доопределить в блоке i-bem__dom на уровне проекта, чтоб не копипастить в каждый блок

**Comment by belozyorcev** [comment]
@Guria, как раз начал задумываться над этим :) А за намёк спасибо )

**Comment by apsavin** [comment]
Ребята, что вы делаете?
Если уж сильно хочется писать `this.elem().setMod` - расширьте jQuery.fn один раз и пишите.

**Comment by Guria** [comment]
Я пока пользуюсь тем что дают и не парюсь :)

**Comment by belozyorcev** [comment]
## Короткая запись модификации элементов
было:  ```this.setMod(this.elem('body'), 'size', 'xxx'); ```
стало: ```this.el('body').setMod('size', 'xxx'); ```

### Реализация
1. в каталоге **common.blocks** создаём каталог **i-bem** в котором создаём каталог **__dom**
2. в каталоге **__dom** создаём файл **i-bem__dom.browser.js** со следующим содержимым

```js
modules.define('i-bem__dom', function(provide, BEMDOM) {

provide(BEMDOM.decl({
    el: function (name) {
        var elem = this.elem(name);
        var self = this;
        elem.setMod = function (mod, val) {
            self.setMod(this, mod, val)
        };
        return elem;
    },
}));

});
```

П.Н. 
спасибо @Guria за подсказку

**Comment by zxqfox** [comment]
@apsavin А как узнать при этом к какому блоку элемент относится?

**Comment by Guria** [comment]
@belozyorcev вроде норм, только тогда надо другими методами приправить

**Comment by zxqfox** [comment]
@Guria @belozyorcev Мне нравится идея дорулить `.setMod` и другие функции в `$.fn`, но я не очень понимаю, как это можно сделать без знаний о блоках, к которым этот элемент относится.

Но если делать через `this.el` — может быть проще один раз в `$.fn` дописать `.setMod` и другие функции, а в `.el` засовывать сам блок в возвращаемый jQuery-chain? Ну и при использовании без блока просто ругаться.

**Comment by Guria** [comment]
Кажись понял. Не нравится то, что в предложенном решении к каждому элементу навешивается свой отдельный метод? Действительно, наверное не очень хорошо.

**Comment by zxqfox** [comment]
@Guria В том числе, но еще и `this.el` можно было бы заменить на `this.elem`.

@tadatuta А как рулятся такие ситуации?

```html
<div class="block">
  <span class="block__elem></span>
  <div class="block">
    <span class="block__elem></span> <!-- этот -->
  </div>
</div>
```

Элемент внутренний будет в обоих блоках виден?

**Comment by belozyorcev** [comment]
@zxqfox - боюсь что через $.fn  решение окажется более громоздким. Вдобавок ко всему нужно следить за кэшем БЭМ.

**Comment by zxqfox** [comment]
@belozyorcev Но и более универсальным ;-) И не будет путаницы между `.el` и `.elem`

**Comment by belozyorcev** [comment]
@zxqfox
> @apsavin А как узнать при этом к какому блоку элемент относится?

у возвращаемого элемента через ```this.elem('name')``` есть свойство ```__bemElemName```, где указано имя блока, к которому принадлежит элемент.

**Comment by belozyorcev** [comment]
@zxqfox - желание было сделать через ```.elem```, дабы реализовать "подготовительный" этап перед bem-core@v3
Но пока это ```.el```.

**Comment by zxqfox** [comment]
Может как-то так?
```js
var origElem = this.elem;
this.elem = function (...) {
  var ret = origElem(...);
  ret.setMod = ...;
  return ret;
};
```

**Comment by sipayRT** [comment]
> А как рулятся такие ситуации?

@zxqfox я думаю, что в таком случае тебе может помочь `this.findElem('elem, true);`. Со `strictMode` `i-bem` не будет искать элемент глубже.

**Comment by aristov** [comment]
В bem-core@v2 есть [экспериментальная фича](https://github.com/bem/bem-core/blob/v2/common.blocks/i-bem/__dom/_elem-instances/i-bem__dom_elem-instances.ru.md), реализующая работу с инстансами элементов. Для ее использования достаточно добавить в зависимости блока следующее:
```js
mustDeps: { block: 'i-bem', elem: 'dom', mods: { 'elem-instances': true } }
```
Тогда можно устанавливать модификатор напрямую у инстанса элемента:
```
this.elemInstance('body')
      .setMod('size', 'xxx')
      .setMod('color', 'gray');
```

Кроме того, можно работать с БЭМ-событиями на элементах, создавать для них декларации и т.д.

**Comment by belozyorcev** [comment]
@aristov ооо :) Правда запись немного длинная получается

**Comment by aristov** [comment]
@belozyorcev да, но для нас читаемость и понятность кода важнее краткости его написания.
